design-proposals/generalize-virt-stack/gen-virt-stack.md

```diff
@@ -0,0 +1,130 @@
+# Overview
+
+This KubeVirt design proposal discusses how KubeVirt can be used to create `libvirt` virtual machines that are backed by diverse hypervisor drivers, such as QEMU/KVM, Xen, VirtualBox, etc. The aim of this proposal is to enumerate the design and implementation choices for enabling this multi-hypervisor support in KubeVirt. 
+
+## Motivation
+
+Although KubeVirt currently relies on libvirt to create and manage virtual machine instances (VMIs), it relies specifically on the QEMU/KVM virtualization stack (VMM and hypervisor) to host the VMI. This limits KubeVirt from being used in settings where a different VMM or hypervisor is used. 
+
+In fact, libvirt itself is flexible enough to support a diverse set of VMMs and hypervisors. The libvirt API delegates its implementation to one or more internal drivers, dependending on the [connection URI](https://libvirt.org/uri.html) passed when initializing the library. The list of currently supported hypervisor drivers in Libvirt are:
+- [LXC - Linux Containers](https://libvirt.org/drvlxc.html)
+- [OpenVZ](https://libvirt.org/drvopenvz.html)
+- [QEMU/KVM/HVF](https://libvirt.org/drvqemu.html)
+- [VirtualBox](https://libvirt.org/drvvbox.html)
+- [VMware ESX](https://libvirt.org/drvesx.html)
+- [VMware Workstation/Player](https://libvirt.org/drvvmware.html)
+- [Xen](https://libvirt.org/drvxen.html)
+- [Microsoft Hyper-V](https://libvirt.org/drvhyperv.html)
+- [Virtuozzo](https://libvirt.org/drvvirtuozzo.html)
+- [Bhyve - The BSD Hypervisor](https://libvirt.org/drvbhyve.html)
+- [Cloud Hypervisor](https://libvirt.org/drvch.html)
+
+There are several parts in the KubeVirt source-code that hard-code the use of the QEMU/KVM hypervisor driver, which prevents the creation of VMIs using another hypervisor driver. Therefore, KubeVirt needs to be updated to introduce the choice of the backend libvirt hypervisor driver to use for creating a given VMI. This would expand the set of scenarios in which KubeVirt can be used.
+
+## Goals
+
+KubeVirt should be able to offer a choice to its users over which libvirt hypervisor-driver they want to use to create their VMI.
```

Comment:
Why?

This section is impotrant: Please provide a justification of how this will help KubeVirt users, or KubeVirt.

Reply:
Microsoft is a vendor of KubeVirt as it leverages KubeVirt in its Azure Operator Nexus product as a VM orchestrator. The hypervisor currently used in the Nexus product is qemu-kvm, however, in the future MSFT is looking at alternative hypervisors such as cloud-hypervisor.
To continue using KubeVirt for this product it would make sense to make it hypervisor-agnostic.

Reply:
There is also another project called Virtink which was created to add K8s-based orchestration support for cloud-hypervisor based VMs.
https://github.com/smartxworks/virtink
This shows that there is a need for K8s-based orchestration for cloud-hypervisor VMs, and KubeVirt already interfaces with libvirt - which supports cloud-hypervisor driver.